src/entities.lisp

```diff
@@ -0,0 +1,96 @@
+;;;; entities.lisp
+
+(in-package #:sketch)
+
+;;;  _____ _   _ _____ ___ _____ ___ _____ ____
+;;; | ____| \ | |_   _|_ _|_   _|_ _| ____/ ___|
+;;; |  _| |  \| | | |  | |  | |  | ||  _| \___ \
+;;; | |___| |\  | | |  | |  | |  | || |___ ___) |
+;;; |_____|_| \_| |_| |___| |_| |___|_____|____/
+;;;
+
+;;; TODO:
+;;; - Better code reuse with sketch.lisp
+
+(defparameter *entity* nil
+  "The current entity.")
+
+(defclass entity ()
+  ((width :initform 100 :accessor entity-width :initarg :width)
+   (height :initform 100 :accessor entity-height :initarg :height)))
+
+(defmethod register-entity ((sketch sketch) (entity entity) box)
+  (setf (gethash entity (sketch-%entities sketch))
+	;; TODO: sb-cga shouldn't be used directly from here.
+	(cons (sb-cga:inverse-matrix (env-model-matrix (sketch-%env sketch))) box)))
+
+(defmethod initialize-instance :after ((instance entity) &rest initargs &key &allow-other-keys)
+  (apply #'prepare instance initargs))
+
+(defmethod update-instance-for-redefined-class :after
+    ((instance entity) added-slots discarded-slots property-list &rest initargs)
+  (declare (ignore added-slots discarded-slots property-list))
+  (apply #'prepare instance initargs))
+
+(defun define-entity-defclass (name bindings)
+  `(defclass ,name (entity)
+     (,@(loop for b in bindings
+              unless (eq 'entity (binding-prefix b))
+		collect `(,(binding-name b)
+                          :initarg ,(binding-initarg b)
+                          :accessor ,(binding-accessor b)
+			  ,@(when (binding-channelp b) '(:allocation :class)))))))
+
+(defun define-entity-channel-observers (entity-name bindings)
+  (loop for b in bindings
+        when (binding-channelp b)
+        collect `(define-channel-observer
+                   (setf (,(binding-accessor b) (default-entity-instance ',entity-name))
+                         (in ,(binding-channel-name b)
+                             ,(binding-initform b))))))
+
+(defun define-entity-draw-method (name bindings body)
+  `(defmethod draw ((*entity* ,name) x y
+		    &key (width (entity-width *entity*)) (height (entity-height *entity*)) mode
+		    &allow-other-keys)
+     (declare (ignore mode))
+     (let ((from-width width)
+	   (from-height height)
+	   (bounds (absolute-bounds x y width height)))
```

Comment:
It might be easier to save width & height only. In this case in `propagate-to-entity` you'd just need to check against transformed vertices (`ix` and `iy` instead of `x` and `y`)

Reply:
Not sure I understand

Reply:
I'll try to open a PR then (and see if that's doable at all and I didn't mix something up)